src/endless/EndlessUsbToolDlg.cpp

```diff
@@ -5323,3 +5333,267 @@ BOOL CEndlessUsbToolDlg::IsBitlockedDrive(const CString &drive)
 
 	return retResult;
 }
+
+CStringW CEndlessUsbToolDlg::GetExePath()
+{
+	wchar_t *exePath = NULL;
+	DWORD neededSize = MAX_PATH;
+	CStringW retValue = L"";
+
+	do {
+		if (exePath != NULL) safe_free(exePath);
+		exePath = (wchar_t*)malloc((neededSize + 1) * sizeof(wchar_t));
+		memset(exePath, 0, (neededSize + 1) * sizeof(wchar_t));
+		DWORD result = GetModuleFileNameW(NULL, exePath, neededSize);
+		IFFALSE_GOTOERROR(result != 0 && (GetLastError() == ERROR_SUCCESS || GetLastError() == ERROR_INSUFFICIENT_BUFFER), "Could not get needed size for module path");
+		neededSize = result;
+		IFFALSE_CONTINUE(GetLastError() != ERROR_INSUFFICIENT_BUFFER, "Not enough memory allocated for buffer. Trying again.");
+		break;
+	} while (TRUE);
+
+	retValue = exePath;
+error:
+	if (exePath != NULL) safe_free(exePath);
+
+	return retValue;
+}
+
+#define REGKEY_WIN_UNINSTALL	L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion\\Uninstall\\"
+#define REGKEY_ENDLESS_OS		ENDLESS_OS_NAME
+#define REGKEY_UNINSTALLENDLESS (REGKEY_WIN_UNINSTALL REGKEY_ENDLESS_OS)
+#define REGKEY_DISPLAYNAME		L"DisplayName"
+#define REGKEY_DISPLAYVERSION	L"DisplayVersion"
+#define REGKEY_UNINSTALL_STRING	L"UninstallString"
+#define REGKEY_INSTALL_LOCATION	L"InstallLocation"
+#define REGKEY_PUBLISHER		L"Publisher"
+#define REGKEY_VERSION_MAJOR	L"VersionMajor"
+#define REGKEY_VERSION_MINOR	L"VersionMinor"
+#define REGKEY_HELP_LINK		L"HelpLink"
+#define REGKEY_NOCHANGE			L"NoChange"
+#define REGKEY_NOMODIFY			L"NoModify"
+
+#define REGKEY_DISPLAYNAME_TEXT	ENDLESS_OS_NAME
+#define REGKEY_PUBLISHER_TEXT	L"Endless Mobile, Inc."
+
+BOOL CEndlessUsbToolDlg::AddUninstallRegistryKeys(const CStringW &uninstallExePath, const CStringW &installPath, DWORD versionMajor, DWORD versionMinor)
+{
+	CRegKey registryKey;
+	LSTATUS result;
+	BOOL retResult = FALSE;
+	CStringW displayVersion;
+
+	result = registryKey.Create(HKEY_LOCAL_MACHINE, REGKEY_UNINSTALLENDLESS);
+	uprintf("Result %d on creating key '%ls'", result, REGKEY_UNINSTALLENDLESS);
+	IFFALSE_GOTOERROR(result == ERROR_SUCCESS, "Error on CRegKey::Create.");
+
+	IFFALSE_GOTOERROR(SetEndlessRegistryKey(HKEY_LOCAL_MACHINE, REGKEY_UNINSTALLENDLESS, REGKEY_DISPLAYNAME, REGKEY_DISPLAYNAME_TEXT, false), "Error on REGKEY_DISPLAYNAME");
+	displayVersion.Format(L"%d.%d", versionMajor, versionMinor);
```

Comment:
This will, in time, become stale: when we release Endless OS 3.1, the installed system will update itself, but this will not be up to date.

I guess this is not a massive problem.